### PR TITLE
Report progress on Dismissible update callback

### LINKED ISSUE #74582: expose the moveController of the Dismissible widget
My proposal is to expose an interface for the Dismissible widget (in the flutter/material.dart package) to show the value of the current offset in which the item has been dragged. 

In my use case I wanted to implement an animation of an object fading in while the dismissible is beeing dismissed, in such a way that the offset of the dismissed item equals the opacity at each moment. I was able to do so only by changing _DismissibleState, _moveController and _dragExtent from private to public. I used flutter stable channel 1.22.5.

My code looks like this:

```
GlobalKey<DismissibleState> currentKey = GlobalKey<DismissibleState>();

...

@override
Widget build(BuildContext context) {
  ...
  Listener(
    onPointerDown: (PointerDownEvent event) {
      currentKey.currentState.moveController.removeListener(dragListener);
      currentKey.currentState.moveController.addListener(dragListener);
    },
    child: Dismissible(
      key: currentKey,
      onDismissed: (direction) {
        currentKey.currentState.moveController.removeListener(dragListener);
        // do some more stuff ...
      },
      child: ...,
      ...
    ),
  ...
}

...

@override
void dispose() {
  currentKey.currentState.moveController.removeListener(dragListener);
  super.dispose();
}

```

where dragListener uses moveController.value and dragExtent.sign to set the state of the faded object.

I tried to do it without changing the flutter base code, by listening to onPointerDown, onPointerUp and onPointerMove events, but then I had to compute myself the offset and it was very buggy because there are many cases were the user can do funny stuff.

I couldn't find any other solution on the internet. I would appriciate if this could become available in future release.

Kind regards

### PR DESCRIPTION
**Pull request by cachapa**
This PR adds a `progress` parameter to the Dismissible `onUpdate` callback.
The value represents how far the dismissible child has been dragged in its parent container as a ratio value between 0.0 and 1.0, which can be useful to e.g. fade elements in sync with the gesture.

This implementation also correctly reports the position during the return animation when the dismiss is unsuccessful.

This PR partly solves #21230 (3rd. point) and closes #74582

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/dismissible.dart [hunk 1]
```diff
@@ -438,6 +444,7 @@ class _DismissibleState extends State<Dismissible> with TickerProviderStateMixin
           direction: _dismissDirection,
           reached: _dismissThresholdReached,
           previousReached: oldDismissThresholdReached,
+          progress: _moveController!.value,
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/dismissible.dart:447]: Should this be the `_moveController!.value` or the progress towards the actual threshold?

File: packages/flutter/test/widgets/dismissible_test.dart [hunk 2]
```diff
@@ -12,6 +12,7 @@ const DismissDirection defaultDismissDirection = DismissDirection.horizontal;
 const double crossAxisEndOffset = 0.5;
 bool reportedDismissUpdateReached = false;
 bool reportedDismissUpdatePreviousReached = false;
+double reportedDismissUpdateProgress = 0.0;
```
Comment (Piinks) [inline packages/flutter/test/widgets/dismissible_test.dart:15]: Can you add tests for non-zero cases?

File: packages/flutter/lib/src/widgets/dismissible.dart (hunk 1 above)
Comment (cachapa) [reply packages/flutter/lib/src/widgets/dismissible.dart:447]: If I understand correctly, the threshold is the value after which the dismiss action completes sucessfully if the pointer is lifted?

The value we want to report in the callback is the actual visibility of the dismissible, regardless of the threshold, since it will be used to synchronize other elements to what the dismissible is doing on screen, e.g. fading the dismissible as it's dragged offscreen - you'd want the opacity to be at 0.0 just as the widget is completely offscreen.

File: packages/flutter/lib/src/widgets/dismissible.dart [hunk 3]
```diff
@@ -247,6 +248,11 @@ class DismissUpdateDetails {
   /// This can be used in conjunction with [DismissUpdateDetails.reached] to catch the moment
   /// that the [Dismissible] is dragged across the threshold.
   final bool previousReached;
+
+  /// The offset ratio of the dismissible in its parent container.
+  /// A value of 0.0 represents the normal position and 1.0 means the child is
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/dismissible.dart:253]: Style guide nit:
```suggestion
  ///
  /// A value of 0.0 represents the normal position and 1.0 means the child is
```

File: packages/flutter/lib/src/widgets/dismissible.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/dismissible.dart:447]: > The value we want to report in the callback is the actual visibility of the dismissible, regardless of the threshold, since it will be used to synchronize other elements to what the dismissible is doing on screen, e.g. fading the dismissible as it's dragged offscreen - you'd want the opacity to be at 0.0 just as the widget is completely offscreen.

Something like this would be excellent to include in the documentation, up around line 250ish :)

File: packages/flutter/test/widgets/dismissible_test.dart (hunk 2 above)
Comment (Piinks) [reply packages/flutter/test/widgets/dismissible_test.dart:15]: Can you add another test that validates it reports the right value somewhere in between 0 and 1? Thanks!

File: packages/flutter/lib/src/widgets/dismissible.dart (hunk 1 above)
Comment (cachapa) [reply packages/flutter/lib/src/widgets/dismissible.dart:451]: Thanks. I've tried to improve it.
As a software developer, I'm chronically incapable of writing good concise documentation so feedback is welcome :-)

File: packages/flutter/test/widgets/dismissible_test.dart (hunk 2 above)
Comment (cachapa) [reply packages/flutter/test/widgets/dismissible_test.dart:15]: Done.

**Review by Piinks — APPROVED** [review]
This LGTM! Thank you for the contribution! 
I've reached out to the team for a secondary review.

**Comment by cachapa** [comment]
Thanks. I'm sorry it took so long, I didn't notice that the tests were failing since last week.

**Review by Hixie — APPROVED** [review]
LGTM

**Comment by cachapa** [comment]
I'm kind of new to this, is there anything else required from my side?

**Comment by Hixie** [comment]
Nope sorry, we just didn't notice it was ready to land! My bad!
